clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -756,19 +737,33 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
                 }
 
                 List<ConsumerRecord<K, V>> parsed = new ArrayList<>();
+                boolean skippedRecords = false;
                 for (LogEntry logEntry : partition.records.deepEntries()) {
                     // Skip the messages earlier than current position.
                     if (logEntry.offset() >= position) {
                         parsed.add(parseRecord(tp, logEntry));
                         bytes += logEntry.sizeInBytes();
-                    }
+                    } else
+                        skippedRecords = true;
                 }
 
                 recordsCount = parsed.size();
 
                 if (!parsed.isEmpty()) {
                     log.trace("Adding fetched record for partition {} with offset {} to buffered record list", tp, position);
                     parsedRecords = new PartitionRecords<>(fetchOffset, tp, parsed);
+                } else if (!skippedRecords && (partition.records.sizeInBytes() > 0)) {
+                    // We did not read a single message from a non-empty buffer, because that message's size is
```

Comment:
Interestingly, [KIP-74](https://cwiki.apache.org/confluence/display/KAFKA/KIP-74%3A+Add+Fetch+Response+Size+Limit+in+Bytes) mentions that "If max_bytes is Int.MAX_INT, new request behaves exactly like old one."  And also that "FetchRequest with top level max_bytes != Int.MAX_INT always makes progress - if server has message(s), than at least one message is returned irrespective of max_bytes."  To me, this implies that if you set max_bytes == Int.MAX_INT, the new broker should give the old behavior where it fails to return any messages, and the client throws the MessageTooLargeException.  Looking at Fetcher.java and ReplicaManager.scala, I can see that this doesn't seem to have been implemented, however.  We should probably clarify whether we will ever implement this behavior before making comments about the version based on the behavior.